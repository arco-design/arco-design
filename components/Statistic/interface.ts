import { ReactNode, CSSProperties } from 'react';
import { Dayjs } from 'dayjs';

/**
 * @title Statistic
 */
export interface StatisticProps {
  style?: CSSProperties;
  className?: string | string[];
  /**
   * @zh 数值的样式
   * @en The css style of statistic's value
   */
  styleValue?: CSSProperties;
  /**
   * @zh 数值小数部分的样式
   * @en The style of the decimal part of the value
   * @version 2.47.0
   */
  styleDecimal?: CSSProperties;
  /**
   * @zh 数值的标题
   * @en The title
   */
  title?: string | ReactNode;
  /**
   * @zh 数值
   * @en Display value
   */
  value?: string | number | Dayjs;
  /**
   * @zh 数字精度
   * @en The precision of the value
   */
  precision?: number;
  /**
   * @zh 显示千位分割符
   * @en Whether to display thousands separator
   */
  groupSeparator?: boolean;
  /**
   * @zh 前缀
   * @en The prefix element
   */
  prefix?: string | ReactNode;
  /**
   * @zh 后缀
   * @en The suffix element
   */
  suffix?: string | ReactNode;
  /**
   * @zh 在数值下展示额外内容
   * @en he extra content to be rendered under the value
   */
  extra?: ReactNode;
  /**
   * @zh 数字动态变大
   * @en Whether the value becomes larger dynamically
   */
  countUp?: boolean;
  /**
   * @zh 从什么数字开始动态变大
   * @en The number that the value starts to increase dynamically
   * @defaultValue 0
   */
  countFrom?: number;
  /**
   * @zh 动态变大的过渡时间 (ms)
   * @en Dynamic time (ms)
   * @defaultValue 2000
   */
  countDuration?: number;
  /**
   * @zh [dayjs](https://github.com/iamkun/dayjs)'s format
   * @en [dayjs](https://github.com/iamkun/dayjs)'s format
   */
  format?: string;
  /**
   * @zh 自定义 render 函数。`formattedValue` 表示格式化后的值。
   * @en Custom render function. `formattedValue` represents the formatted value.
   * @version 2.36.0
   */
  renderFormat?: (value: StatisticProps['value'], formattedValue: string) => ReactNode;
  /**
   * @zh 数字是否加载中
   * @en Is the number loading
   * @version 2.20.0
   */
  loading?: boolean;
}

/**
 * @title Statistic.Countdown
 */
export interface CountdownProps {
  style?: CSSProperties;
  className?: string | string[];
  /**
   * @zh 数值的样式
   * @en The css style of statistic's value
   */
  styleValue?: CSSProperties;
  /**
   * @zh 数值的标题
   * @en The title element
   */
  title?: string | ReactNode;
  /**
   * @zh 倒计时的时间
   * @en To set value
   */
  value?: number | string | Date | Dayjs;
  /**
   * @zh [dayjs](https://github.com/iamkun/dayjs)'s format
   * @en [dayjs](https://github.com/iamkun/dayjs)'s format
   * @defaultValue HH:mm:ss
   */
  format?: string;
  /**
   * @zh 自定义 render 函数。`valueDiff` 表示两个时间的时间差，`formattedDiff` 表示格式化后的时间差，
   * @en Custom render function, the input parameter is the result formatted by `dayjs`
   * @version 2.36.0
   */
  renderFormat?: (valueDiff: number, formattedDiff: string) => ReactNode;
  /**
   * @zh 倒计时完成后触发的回调
   * @en Callback at the end of the countdown
   */
  onFinish?: () => void;
  /**
   * @zh 是否开始倒计时，默认为 `true`，可以通过设置该值控制倒计时的时机
   * @en Whether to start the countdown
   * @defaultValue true
   */
  start?: boolean;
  /**
   * @zh 用于修正初始化时间显示不正确
   * @en The current time. Used to correct the initialization time
   */
  now?: number | string | Date | Dayjs;
}
