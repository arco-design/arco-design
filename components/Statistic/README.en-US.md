`````
Component / Data Display

# Statistic

Display statistic number.
`````

%%Content%%

## API

### Statistic

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|countUp|Whether the value becomes larger dynamically|boolean |`-`|-|
|groupSeparator|Whether to display thousands separator|boolean |`-`|-|
|loading|Is the number loading|boolean |`-`|2.20.0|
|countDuration|Dynamic time (ms)|number |`2000`|-|
|countFrom|The number that the value starts to increase dynamically|number |`0`|-|
|precision|The precision of the value|number |`-`|-|
|format|[dayjs](https://github.com/iamkun/dayjs)'s format|string |`-`|-|
|renderFormat|Custom render function. `formattedValue` represents the formatted value.|(value: [StatisticProps](statistic#statistic)['value'], formattedValue: string) => ReactNode |`-`|2.36.0|
|extra|he extra content to be rendered under the value|ReactNode |`-`|-|
|prefix|The prefix element|string \| ReactNode |`-`|-|
|suffix|The suffix element|string \| ReactNode |`-`|-|
|title|The title|string \| ReactNode |`-`|-|
|className|Additional css class|string \| string[] |`-`|-|
|style|Additional style|CSSProperties |`-`|-|
|styleValue|The css style of statistic's value|CSSProperties |`-`|-|
|value|Display value|string \| number \| Dayjs |`-`|-|

### Statistic.Countdown

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|start|Whether to start the countdown|boolean |`true`|-|
|format|[dayjs](https://github.com/iamkun/dayjs)'s format|string |`HH:mm:ss`|-|
|title|The title element|string \| ReactNode |`-`|-|
|className|Additional css class|string \| string[] |`-`|-|
|now|The current time. Used to correct the initialization time|number \| string \| Date \| Dayjs |`-`|-|
|style|Additional style|CSSProperties |`-`|-|
|styleValue|The css style of statistic's value|CSSProperties |`-`|-|
|value|To set value|number \| string \| Date \| Dayjs |`-`|-|
|onFinish|Callback at the end of the countdown|() => void |`-`|-|
|renderFormat|Custom render function, the input parameter is the result formatted by `dayjs`|(valueDiff: number, formattedDiff: string) => ReactNode |`-`|2.36.0|
