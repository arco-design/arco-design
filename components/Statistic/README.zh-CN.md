`````
组件 / 数据展示

# 数值显示 Statistic

突出展示某个或某组数字、带描述的统计类数据。
`````

%%Content%%

## API
### Statistic

|参数名|描述|类型|默认值|版本|
|---|---|---|---|---|
|countUp|数字动态变大|boolean |`-`|-|
|groupSeparator|显示千位分割符|boolean |`-`|-|
|loading|数字是否加载中|boolean |`-`|2.20.0|
|countDuration|动态变大的过渡时间 (ms)|number |`2000`|-|
|countFrom|从什么数字开始动态变大|number |`0`|-|
|precision|数字精度|number |`-`|-|
|format|[dayjs](https://github.com/iamkun/dayjs)'s format|string |`-`|-|
|renderFormat|自定义 render 函数。`formattedValue` 表示格式化后的值。|(value: [StatisticProps](statistic#statistic)['value'], formattedValue: string) => ReactNode |`-`|2.36.0|
|extra|在数值下展示额外内容|ReactNode |`-`|-|
|prefix|前缀|string \| ReactNode |`-`|-|
|suffix|后缀|string \| ReactNode |`-`|-|
|title|数值的标题|string \| ReactNode |`-`|-|
|className|节点类名|string \| string[] |`-`|-|
|style|节点样式|CSSProperties |`-`|-|
|styleValue|数值的样式|CSSProperties |`-`|-|
|value|数值|string \| number \| Dayjs |`-`|-|

### Statistic.Countdown

|参数名|描述|类型|默认值|版本|
|---|---|---|---|---|
|start|是否开始倒计时，默认为 `true`，可以通过设置该值控制倒计时的时机|boolean |`true`|-|
|format|[dayjs](https://github.com/iamkun/dayjs)'s format|string |`HH:mm:ss`|-|
|title|数值的标题|string \| ReactNode |`-`|-|
|className|节点类名|string \| string[] |`-`|-|
|now|用于修正初始化时间显示不正确|number \| string \| Date \| Dayjs |`-`|-|
|style|节点样式|CSSProperties |`-`|-|
|styleValue|数值的样式|CSSProperties |`-`|-|
|value|倒计时的时间|number \| string \| Date \| Dayjs |`-`|-|
|onFinish|倒计时完成后触发的回调|() => void |`-`|-|
|renderFormat|自定义 render 函数。`valueDiff` 表示两个时间的时间差，`formattedDiff` 表示格式化后的时间差，|(valueDiff: number, formattedDiff: string) => ReactNode |`-`|2.36.0|
