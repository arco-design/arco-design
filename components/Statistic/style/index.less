@import './token.less';

@statistic-prefix-cls: ~'@{prefix}-statistic';

.@{statistic-prefix-cls} {
  display: inline-block;
  line-height: @line-height-base;
  color: @statistic-color-text;

  &-title {
    font-size: @statistic-font-title-size;
    margin-bottom: @statistic-margin-title-bottom;
    color: @statistic-color-title-text;
  }

  &-content {
    .@{statistic-prefix-cls}-value {
      white-space: nowrap;
      font-size: @statistic-font-int-size;
      font-weight: @statistic-font-value-weight;
      color: @statistic-color-value-text;

      &-int {
        white-space: nowrap;
      }

      &-decimal {
        font-size: @statistic-font-decimal-size;
        display: inline-block;
      }

      &-suffix {
        font-size: @statistic-font-suffix-size;
        margin-left: @statistic-margin-suffix-left;
      }

      &-prefix {
        font-size: @statistic-size-value-icon;
        margin-right: @statistic-margin-prefix-right;
      }
    }
  }

  &-extra {
    margin-top: @statistic-margin-extra-top;
    color: @statistic-color-extra-text;
  }
}

.@{statistic-prefix-cls}-rtl {
  .@{statistic-prefix-cls} {
    &-content {
      .@{statistic-prefix-cls}-value {
        &-suffix {
          margin-right: @statistic-margin-suffix-left;
          margin-left: 0;
        }

        &-prefix {
          margin-right: 0;
          margin-left: @statistic-margin-prefix-right;
        }
      }
    }
  }
}
