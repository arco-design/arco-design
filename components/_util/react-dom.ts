import { ReactElement } from 'react';
import ReactDOM from 'react-dom';
import { isObject } from './is';

type CreateRootFnType = (container: Element | DocumentFragment) => {
  render: (container: ReactElement) => void;
  unmount: () => void;
  _unmount: () => void;
};

const __SECRET_INTERNALS__ = '__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED';

const CopyReactDOM = ReactDOM as typeof ReactDOM & {
  createRoot: CreateRootFnType;
  // https://github.com/facebook/react/blob/4ff5f5719b348d9d8db14aaa49a48532defb4ab7/packages/react-dom/src/client/ReactDOM.js#L181
  [__SECRET_INTERNALS__]: {
    usingClientEntryPoint?: boolean;
  };
};

let copyRender: (
  app: ReactElement,
  container: Element | DocumentFragment
) => {
  render: (container: ReactElement) => void;
  _unmount: () => void;
};

const isReact18 = Number(CopyReactDOM.version?.split('.')[0]) > 17;

const updateUsingClientEntryPoint = (skipWarning?: boolean) => {
  // https://github.com/facebook/react/blob/17806594cc28284fe195f918e8d77de3516848ec/packages/react-dom/npm/client.js#L10
  // Avoid console warning
  if (isObject(CopyReactDOM[__SECRET_INTERNALS__])) {
    CopyReactDOM[__SECRET_INTERNALS__].usingClientEntryPoint = skipWarning;
  }
};

let createRoot: CreateRootFnType;
try {
  createRoot = CopyReactDOM.createRoot;
} catch (_) {
  //
}

if (isReact18 && createRoot) {
  copyRender = (app: ReactElement, container: Element | DocumentFragment) => {
    updateUsingClientEntryPoint(true);
    const root = createRoot(container);
    updateUsingClientEntryPoint(false);

    root.render(app);

    root._unmount = function () {
      setTimeout(() => {
        root?.unmount?.();
      });
    };
    return root;
  };
} else {
  copyRender = function (app: ReactElement, container: Element | DocumentFragment) {
    CopyReactDOM.render(app, container);

    return {
      render: (app: ReactElement) => {
        CopyReactDOM.render(app, container);
      },
      _unmount() {
        CopyReactDOM.unmountComponentAtNode(container);
      },
    };
  };
}

export const render = copyRender;
