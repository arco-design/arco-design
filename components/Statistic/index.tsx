import React, {
  useState,
  useEffect,
  useContext,
  forwardRef,
  useRef,
  useImperativeHandle,
  useMemo,
} from 'react';
import BTween from 'b-tween';
import dayjs, { Dayjs } from 'dayjs';
import cs from '../_util/classNames';
import Countdown from './countdown';
import { isNumber } from '../_util/is';
import { ConfigContext } from '../ConfigProvider';
import Skeleton from '../Skeleton';
import { StatisticProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';

type StatisticHandle = {
  countUp: () => void;
};

const defaultProps: StatisticProps = {
  countFrom: 0,
  countDuration: 2000,
};

function Statistic(baseProps: StatisticProps, ref) {
  const { getPrefixCls, componentConfig } = useContext(ConfigContext);
  const props = useMergeProps<StatisticProps>(baseProps, defaultProps, componentConfig?.Statistic);
  const {
    className,
    style,
    title,
    extra,
    groupSeparator,
    precision,
    prefix,
    suffix,
    format,
    styleValue,
    loading,
  } = props;

  const tween = useRef<BTween>();
  const [value, setValue] = useState<string | number | Dayjs>(
    'value' in props ? props.value : undefined
  );

  const prefixCls = getPrefixCls('statistic');

  const countUp = (from = props.countFrom, to = props.value) => {
    const { countDuration } = props;
    if (from !== to) {
      tween.current = new BTween({
        from: {
          value: from,
        },
        to: {
          value: to,
        },
        duration: countDuration,
        easing: 'quartOut',
        onUpdate: (keys) => {
          setValue(keys.value.toFixed(precision));
        },
        onFinish: () => {
          setValue(to);
        },
      });
      tween.current.start();
    }
  };

  useEffect(() => {
    if (props.countUp) {
      if (tween.current) {
        tween.current.stop();
      }
      if (value !== props.value) {
        countUp(Number(value), props.value);
      } else {
        countUp();
      }
    } else {
      setValue(props.value);
    }

    return () => {
      tween.current && tween.current.stop();
      tween.current = null;
    };
  }, [props.value]);

  useImperativeHandle<any, StatisticHandle>(ref, () => ({
    countUp,
  }));

  const { int, decimal } = useMemo(() => {
    let _value = value;
    if (format) {
      _value = dayjs(value).format(format);
    }
    if (isNumber(precision) && precision >= 0) {
      _value = Number(value).toFixed(precision);
    }
    let int = String(_value).split('.')[0];
    const decimal = String(_value).split('.')[1];
    if (groupSeparator && isNumber(Number(value))) {
      int = Number(int).toLocaleString('en-US');
    }
    return {
      int,
      decimal,
    };
  }, [format, groupSeparator, precision, value]);

  return (
    <div className={cs(`${prefixCls}`, className)} style={style}>
      {title && <div className={`${prefixCls}-title`}>{title}</div>}
      <div className={`${prefixCls}-content`}>
        <Skeleton animation loading={!!loading} text={{ rows: 1, width: '100%' }}>
          <div className={`${prefixCls}-value`} style={styleValue}>
            {!isNumber(Number(value)) ? (
              value
            ) : (
              <span className={`${prefixCls}-value-int`}>
                <span className={`${prefixCls}-value-prefix`}>{prefix}</span> {int}
              </span>
            )}

            {decimal !== undefined || suffix ? (
              <span className={`${prefixCls}-value-decimal`}>
                {isNumber(Number(value)) && decimal !== undefined && `.${decimal}`}{' '}
                {suffix && <span className={`${prefixCls}-value-suffix`}>{suffix}</span>}
              </span>
            ) : null}
          </div>
        </Skeleton>
        {extra && <div className={`${prefixCls}-extra`}>{extra}</div>}
      </div>
    </div>
  );
}

const ForwardRefStatistic = forwardRef(Statistic);

const StatisticComponent = ForwardRefStatistic as typeof ForwardRefStatistic & {
  Countdown: typeof Countdown;
};

StatisticComponent.displayName = 'Statistic';

StatisticComponent.Countdown = Countdown;

export default StatisticComponent;

export { StatisticProps };
