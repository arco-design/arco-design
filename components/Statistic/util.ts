import { padStart } from '../_util/pad';

const units: [string, number][] = [
  ['Y', 1000 * 60 * 60 * 24 * 365], // years
  ['M', 1000 * 60 * 60 * 24 * 30], // months
  ['D', 1000 * 60 * 60 * 24], // days
  ['H', 1000 * 60 * 60], // hours
  ['m', 1000 * 60], // minutes
  ['s', 1000], // seconds
  ['S', 1], // million seconds
];

export function getDateString(millisecond: number, format: string) {
  let leftMillisecond: number = millisecond;
  return units.reduce((current, [name, unit]) => {
    if (current.indexOf(name) !== -1) {
      const value = Math.floor(leftMillisecond / unit);
      leftMillisecond -= value * unit;
      return current.replace(new RegExp(`${name}+`, 'g'), (match: string) => {
        const len = match.length;
        return padStart(value.toString(), len, '0');
      });
    }
    return current;
  }, format);
}
