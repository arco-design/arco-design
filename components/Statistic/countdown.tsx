import React, { forwardRef, useState, useEffect, useContext, useRef } from 'react';
import dayjs, { Dayjs } from 'dayjs';
import { getDayjsValue, getNow } from '../_util/dayjs';
import cs from '../_util/classNames';
import { getDateString } from './util';
import { ConfigContext } from '../ConfigProvider';
import { CountdownProps } from './interface';
import { isFunction } from '../_util/is';

function Countdown(props: CountdownProps, ref) {
  const { getPrefixCls } = useContext(ConfigContext);
  const {
    className,
    style,
    title,
    styleValue,
    value,
    onFinish,
    renderFormat,
    format = 'HH:mm:ss',
    start = true,
  } = props;

  const dayjsValue = (getDayjsValue(value, format) as Dayjs) || dayjs();
  const now = getDayjsValue(props.now, format) as Dayjs;

  const prefixCls = getPrefixCls('statistic');

  const [valueDiff, setValueDiff] = useState(dayjsValue.diff(now, 'millisecond'));
  const [valueShow, setValueShow] = useState(getDateString(Math.max(valueDiff, 0), format));
  const timerRef = useRef(null);

  const stopTimer = () => {
    clearInterval(timerRef.current);
    timerRef.current = null;
  };

  const startTimer = () => {
    timerRef.current = setInterval(() => {
      const _valueDiff = dayjsValue.diff(getNow());
      const _value = dayjsValue.diff(getNow(), 'millisecond');
      if (_value <= 0) {
        stopTimer();
        onFinish && onFinish();
      }
      const valueShow = getDateString(Math.max(_value, 0), format as string);
      setValueShow(valueShow);
      setValueDiff(_valueDiff);
    }, 1000 / 30);
  };

  useEffect(() => {
    if (!timerRef.current && start) {
      if (dayjsValue.valueOf() >= Date.now()) {
        startTimer();
      }
    }
    return () => {
      stopTimer();
    };
  }, [props.start]);

  const valueShowNode = isFunction(renderFormat) ? renderFormat(valueDiff, valueShow) : valueShow;
  return (
    <div
      ref={ref}
      className={cs(`${prefixCls}`, `${prefixCls}-countdown`, className)}
      style={style}
    >
      {title && <div className={`${prefixCls}-title`}>{title}</div>}
      <div className={`${prefixCls}-content`}>
        <div className={`${prefixCls}-value`} style={styleValue}>
          {valueShowNode}
        </div>
      </div>
    </div>
  );
}

const CountdownComponent = forwardRef<unknown, CountdownProps>(Countdown);

CountdownComponent.displayName = 'StatisticCountdown';

export default CountdownComponent;

export { CountdownProps };
