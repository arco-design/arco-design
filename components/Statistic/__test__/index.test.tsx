import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import Statistic from '..';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import { sleep } from '../../../tests/util';

mountTest(Statistic);
componentConfigTest(Statistic, 'Statistic');

describe('Statistic', () => {
  it('groupSeparator', () => {
    const component = mount(<Statistic groupSeparator value={100000} />);

    expect(component.find('.arco-statistic-value > .arco-statistic-value-int').text()).toBe(
      '100,000'
    );

    act(() => {
      component.setProps({ value: 200000 });
    });

    expect(component.find('.arco-statistic-value > .arco-statistic-value-int').text()).toBe(
      '200,000'
    );
  });

  it('prefix / suffix', () => {
    const component = mount(<Statistic title="Title" value={50} prefix="Prefix" suffix="%" />);

    expect(component.find('.arco-statistic-value > .arco-statistic-value-int').text()).toBe(
      'Prefix50'
    );
    expect(component.find('.arco-statistic-value > .arco-statistic-value-decimal').text()).toBe(
      '%'
    );
  });

  it('precision', () => {
    const component = mount(
      <Statistic title="Title" value={1000000} groupSeparator precision={2} suffix="$" />
    );

    expect(component.find('.arco-statistic-value > .arco-statistic-value-int').text()).toBe(
      '1,000,000'
    );
    expect(component.find('.arco-statistic-value > .arco-statistic-value-decimal').text()).toBe(
      '.00$'
    );
  });

  it('date', () => {
    const component = mount(<Statistic value={1554869813383} format="YYYY/MM/DD HH:mm:ss" />);

    expect(component.find('.arco-statistic-value > .arco-statistic-value-int').text()).toBe(
      '2019/04/10 12:16:53'
    );
  });

  it('countUp', async () => {
    const component = mount(<Statistic title="Title" value={100} countDuration={300} countUp />);

    await act(async () => {
      await sleep(100);
    });

    expect(
      Number(component.find('.arco-statistic-value > .arco-statistic-value-int').text())
    ).toBeGreaterThan(0);
    expect(
      Number(component.find('.arco-statistic-value > .arco-statistic-value-int').text())
    ).toBeLessThan(100);

    await act(async () => {
      await sleep(300);
    });

    expect(
      Number(component.find('.arco-statistic-value > .arco-statistic-value-int').text())
    ).toEqual(100);

    await act(async () => {
      component.setProps({ value: 200 });
      await sleep(100);
    });

    expect(
      Number(component.find('.arco-statistic-value > .arco-statistic-value-int').text())
    ).toBeGreaterThan(100);
    expect(
      Number(component.find('.arco-statistic-value > .arco-statistic-value-int').text())
    ).toBeLessThan(200);

    await act(async () => {
      await sleep(300);
    });

    expect(
      Number(component.find('.arco-statistic-value > .arco-statistic-value-int').text())
    ).toEqual(200);
  });

  it('loading', async () => {
    class LoadingStatistic extends React.Component<{}, { loading: boolean }> {
      state = {
        loading: true,
      };

      render() {
        return <Statistic title="Title" loading={this.state.loading} value={200} />;
      }
    }
    const component = mount(<LoadingStatistic />);

    await act(async () => {
      await sleep(100);
    });

    expect(component.find('.arco-statistic-value > .arco-statistic-value-int').length).toBe(0);
    expect(Number(component.find('.arco-skeleton').length)).toBe(1);

    // update loading
    await act(async () => {
      component.setState({ loading: false });
      await sleep(300);
    });

    expect(
      Number(component.find('.arco-statistic-value > .arco-statistic-value-int').text())
    ).toEqual(200);
    expect(Number(component.find('.arco-skeleton').length)).toBe(0);
  });
});
