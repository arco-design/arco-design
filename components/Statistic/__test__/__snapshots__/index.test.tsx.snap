// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Statistic default 1`] = `
<div
  class="arco-statistic"
>
  <div
    class="arco-statistic-content"
  >
    <div
      class="arco-statistic-value"
    />
  </div>
</div>
`;

exports[`ConfigProvider componentConfig.Statistic set className globally 1`] = `
<div
  class="arco-statistic global-default-class"
>
  <div
    class="arco-statistic-content"
  >
    <div
      class="arco-statistic-value"
    />
  </div>
</div>
`;

exports[`ConfigProvider componentConfig.Statistic set className globally and component set className itself. 1`] = `
<div
  class="arco-statistic component-class"
>
  <div
    class="arco-statistic-content"
  >
    <div
      class="arco-statistic-value"
    />
  </div>
</div>
`;

exports[`ConfigProvider componentConfig.Statistic set data-* property 1`] = `
<div
  class="arco-statistic"
  data-test-name="Statistic"
>
  <div
    class="arco-statistic-content"
  >
    <div
      class="arco-statistic-value"
    />
  </div>
</div>
`;
